src/v/datalake/record_multiplexer.cc

```diff
@@ -0,0 +1,77 @@
+/*
+ * Copyright 2024 Redpanda Data, Inc.
+ *
+ * Licensed as a Redpanda Enterprise file under the Redpanda Community
+ * License (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ * https://github.com/redpanda-data/redpanda/blob/master/licenses/rcl.md
+ */
+#include "datalake/record_multiplexer.h"
+
+#include "iceberg/values.h"
+#include "model/record.h"
+
+namespace datalake {
+record_multiplexer::record_multiplexer(
+  std::unique_ptr<data_writer_factory> writer_factory)
+  : _translator{schemaless_translator()}
+  , _writer_factory{std::move(writer_factory)} {}
+
+ss::future<ss::stop_iteration>
+record_multiplexer::operator()(model::record_batch batch) {
+    if (batch.compressed()) {
+        batch = co_await storage::internal::decompress_batch(std::move(batch));
+    }
+    batch.for_each_record([&batch, this](model::record&& record) {
+        iobuf key = record.release_key();
+        iobuf val = record.release_value();
+        int64_t timestamp = batch.header().first_timestamp.value()
+                            + record.timestamp_delta();
+        int64_t offset = static_cast<int64_t>(batch.base_offset())
+                         + record.offset_delta();
+        int64_t estimated_size = key.size_bytes() + val.size_bytes() + 16;
+
+        // Translate the record
+        auto& translator = get_translator();
+        iceberg::struct_value data = std::visit(
+          [&key, &val, timestamp, offset](schemaless_translator& tr) {
+              return tr.translate_event(
+                std::move(key), std::move(val), timestamp, offset);
+          },
+          translator);
+
+        // Send it to the writer
+        ss::shared_ptr<data_writer> writer = get_writer();
+        writer->add_data_struct(std::move(data), estimated_size);
+    });
+    co_return ss::stop_iteration::no;
+}
+
+ss::future<fragmented_vector<std::unique_ptr<data_writer_result>>>
+record_multiplexer::end_of_stream() {
+    // TODO: once we have multiple _writers this should be a loop
+    if (_writer) {
+        fragmented_vector<std::unique_ptr<data_writer_result>> ret;
+        auto res = _writer->finish();
+        ret.push_back(std::move(res));
+        co_return ret;
+    } else {
+        co_return fragmented_vector<std::unique_ptr<data_writer_result>>{};
+    }
+}
+
+record_multiplexer::translator& record_multiplexer::get_translator() {
+    return _translator;
+}
+
+ss::shared_ptr<data_writer> record_multiplexer::get_writer() {
```

Comment:
is shared ownership necessary? it's not very common in the code base, but also, "shared writer" makes it sound less likely to be necessary.